### PR TITLE
Top miner

### PR DESCRIPTION
**Pull request by DaloroAT**

### CODE REVIEW

File: oml/miners/inbatch_top_pn.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+from oml.interfaces.miners import ITripletsMinerInBatch, TTripletsIds
+from oml.utils.misc_torch import pairwise_dist
+
+
+class TopPNTripletsMiner(ITripletsMinerInBatch):
+    """
+    This miner selects top hard triplets based on distances between features:
+
+    - hard `positive` samples has large distance to the anchor sample
```
Comment (AlekseySh) [inline oml/miners/inbatch_top_pn.py:15]: have

File: oml/miners/inbatch_top_pn.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    - hard `negative` samples has small distance to the anchor sample
```
Comment (AlekseySh) [inline oml/miners/inbatch_top_pn.py:17]: have

File: oml/miners/inbatch_top_pn.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+import numpy as np
+import torch
+from torch import Tensor
+
+from oml.interfaces.miners import ITripletsMinerInBatch, TTripletsIds
+from oml.utils.misc_torch import pairwise_dist
+
+
+class TopPNTripletsMiner(ITripletsMinerInBatch):
```
Comment (AlekseySh) [inline oml/miners/inbatch_top_pn.py:11]: what about `NHardestTripletsMiner`name?

File: oml/miners/inbatch_top_pn.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+    Toward the end of the training, annotation errors can affect final metric. If you are not sure about
+    the quality of your dataset, you can use range instead of integer value for paramters and exclude combinations
+    with the largest distances. For example instead picking `top5` positive examples, you can use `top2 - top5` examples
+
+    """
+
+    def __init__(
+        self,
+        top_positive: Union[Tuple[int, int], List[int], int] = 1,
```
Comment (AlekseySh) [inline oml/miners/inbatch_top_pn.py:27]: if top == 1, we have exactly hard miner, right?

File: oml/miners/inbatch_top_pn.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        """
+
+        self.top_positive_slice = slice(*self._parse_input_arg(top_positive))
+        self.top_negative_slice = slice(*self._parse_input_arg(top_negative))
+
+    @staticmethod
+    def _parse_input_arg(top: Union[Tuple[int, int], List[int], int]) -> List[int]:
+        available_types = (list, tuple, int)
+        assert isinstance(top, available_types), f"Unsupported type of argument, must be only {available_types}"
```
Comment (AlekseySh) [inline oml/miners/inbatch_top_pn.py:44]: let's combine this assrt with `if-else` statement and make it `if-elif-else`

File: oml/miners/inbatch_top_pn.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    Toward the end of the training, annotation errors can affect final metric. If you are not sure about
+    the quality of your dataset, you can use range instead of integer value for paramters and exclude combinations
+    with the largest distances. For example instead picking `top5` positive examples, you can use `top2 - top5` examples
+
+    """
+
+    def __init__(
+        self,
+        top_positive: Union[Tuple[int, int], List[int], int] = 1,
+        top_negative: Union[Tuple[int, int], List[int], int] = 1,
```
Comment (AlekseySh) [inline oml/miners/inbatch_top_pn.py:28]: we need assert that tops have to be more or equal than 1

**Review by AlekseySh — COMMENTED** [review]
1. Let's add test the new miner in the case of (top_p=inf, top_n=inf) has the same behavior as the existing AllTripletsMiner 
2. Let's add n duplicated feature vectors to the batch and check that we observe n groups of triplets 

Add motivation message to All and Hard miners

File: oml/miners/inbatch_top_pn.py (hunk 6 above)
Comment (AlekseySh) [reply oml/miners/inbatch_top_pn.py:28]: n_positive, n_negative

File: oml/miners/inbatch_top_pn.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        )
+
+        return ids_anchor, ids_pos, ids_neg
+
+    def _sample_from_distmat(
+        self,
+        distmat: Tensor,
+        labels: List[int],
+        *_: Any,
+        ignore_anchor_mask: Optional[Union[List[int], Tensor, np.ndarray]] = None,
```
Comment (AlekseySh) [inline oml/miners/inbatch_top_pn.py:79]: doc string

File: tests/test_oml/test_miners/test_inbatch_top_pn.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
+from collections import defaultdict
```
Comment (AlekseySh) [inline tests/test_oml/test_miners/test_inbatch_top_pn.py:1]: add test for `anchor_mask_ignore`

File: oml/miners/inbatch_top_pn.py (hunk 6 above)
Comment (DaloroAT) [reply oml/miners/inbatch_top_pn.py:28]: in parser

File: oml/miners/inbatch_top_pn.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+            ignore_anchor_mask = torch.zeros(len(distmat), dtype=torch.bool, device=distmat.device)
+
+        all_ids_reduced = torch.arange(len(labels), device=distmat.device)[torch.logical_not(ignore_anchor_mask)]
+
+        distmat_reduced = distmat[torch.logical_not(ignore_anchor_mask)]
+        ii_arange = (
+            torch.arange(len(distmat_reduced), device=distmat.device).unsqueeze(-1).expand(distmat_reduced.shape)
+        )
+
+        _, ids_highest_distance = torch.topk(distmat_reduced, k=distmat_reduced.shape[-1], largest=True)
```
Comment (AlekseySh) [inline oml/miners/inbatch_top_pn.py:110]: discuss

File: oml/registry/miners.py [hunk 10]
```diff
@@ -5,13 +5,15 @@
...
 from oml.miners.inbatch_hard_tri import HardTripletsMiner
+from oml.miners.inbatch_top_pn import TopPNTripletsMiner
 from oml.utils.misc import TCfg, dictconfig_to_dict
 
 MINERS_REGISTRY = {
     "all_triplets": AllTripletsMiner,
     "hard_cluster": HardClusterMiner,
     "hard_triplets": HardTripletsMiner,
     "triplets_with_memory": TripletMinerWithMemory,
+    "top_pn": TopPNTripletsMiner,
```
Comment (AlekseySh) [inline oml/registry/miners.py:16]: discuss

File: tests/test_oml/test_miners/test_inbatch_top_pn.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+    for anch, pos, neg in triplets:
+        ids_anchor2positives[anch].add(int(pos))
+        ids_anchor2negatives[anch].add(int(neg))
+
+    for idx_anch, miner_positives in ids_anchor2positives.items():
+        assert len(miner_positives) > 0
+        distmat_ = distmat[idx_anch].clone()
+        # we need to select largest distances across same labels and ignore self distance
+        distmat_[labels != labels[idx_anch]] = -1
```
Comment (AlekseySh) [inline tests/test_oml/test_miners/test_inbatch_top_pn.py:120]: let's use `-float(inf)`

**Review by AlekseySh — COMMENTED** [review]
Discussed offline:
* rename the new miner to HardTripletsMiner (rename its arguments from `top_positive` to `n_positive`)
* than old one would be: HardestTripletsMiner, and make it deprecated

File: oml/miners/inbatch_nhard_tri.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+import torch
+from torch import Tensor
+
+from oml.interfaces.miners import ITripletsMinerInBatch, TTripletsIds
+from oml.utils.misc_torch import pairwise_dist
+
+
+class NHardTripletsMiner(ITripletsMinerInBatch):
+    """
+    This miner selects top hard triplets based on distances between features:
```
Comment (AlekseySh) [inline oml/miners/inbatch_nhard_tri.py:13]: top hard -> n hardest

File: oml/miners/inbatch_nhard_tri.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+    - hard `positive` samples have large distance to the anchor sample
+
+    - hard `negative` samples have small distance to the anchor sample
+
+    Toward the end of the training, annotation errors can affect final metric. If you are not sure about the quality of
+    your dataset, you can use range instead of integer value for paramters and exclude combinations with the largest
+    distances. For example instead picking `top5` positive examples, you can use `top2 - top5` examples.
```
Comment (AlekseySh) [inline oml/miners/inbatch_nhard_tri.py:21]: `top5` -> 5 
`top2 - top5` -> you can use examples from the 2nd hardest to the 5th one.

File: tests/test_oml/test_miners/test_inbatch_nhard_tri.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+    expected_anchors = (ignore_anchor_mask == 0).nonzero().squeeze().tolist()
+    assert set(ids_a) == set(expected_anchors)
+
+    check_triplets_are_top(distmat=distmat, labels=labels, nhard_miner=nhard_miner, triplets=triplets)
+
+    check_triplets_consistency(ids_anchor=ids_a, ids_pos=ids_p, ids_neg=ids_n, labels=labels)
+
+
+def check_triplets_are_top(
```
Comment (AlekseySh) [inline tests/test_oml/test_miners/test_inbatch_nhard_tri.py:106]: same rename

**Review by AlekseySh — COMMENTED** [review]
minor renames

**Review by AlekseySh — APPROVED** [review]

**Pull request merged**